### PR TITLE
Feature/cu kk776f ability activation effect

### PR DESCRIPTION
**Pull request by devildrake**
Inside scene HudEffectsWithAbilityFX a UI can be found with both the health features and the ability CD working (testeable on dash, which is the only ability currently working)

### CODE REVIEW

**Comment by jcarlos0305** [comment]
Task linked: [CU-kk776f Ability activation effect](https://app.clickup.com/t/kk776f)

**Comment by jcarlos0305** [comment]
Task linked: [CU-kk6zfz VS3 UI Features](https://app.clickup.com/t/kk6zfz)

File: Penteract/Assets/Scripts/AbilityRefreshEffect.cpp [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+		isPlaying = false;
+		effectMember1->Disable();
+	}
+}
+
+void AbilityRefreshEffect::Play() {
+	if (!skillObjTransform2D) return;
+	if (effectMember1) {
+		effectMember1->Enable();
+		bool active = effectMember1->IsActive();
```
Comment (GBMiro) [inline Penteract/Assets/Scripts/AbilityRefreshEffect.cpp:66]: Variable in line 66 is not being used. You should remove it

File: Penteract/Assets/Scripts/HUDController.cpp [hunk 2]
```diff
@@ -359,23 +379,29 @@ void HUDController::UpdateCommonSkill() {
...
-			ComponentImage* image = (*itSkills)->GetComponent<ComponentImage>();
+	for (std::vector<GameObject*>::iterator it = skills.begin(); it != skills.end(); ++it) {
+
+		if ((*it)->GetChildren().size() > 0) {
+			ComponentImage* image = (*it)->GetChildren()[1]->GetComponent<ComponentImage>();
 			if (image) {
 				if (image->IsFill()) {
 					switch (skill) {
 					case Cooldowns::FANG_SKILL_1:
+						if (isMain) {
```
Comment (GBMiro) [inline Penteract/Assets/Scripts/HUDController.cpp:394]: Either remove these brackets or add them in the others switch cases to keep consistency

**Review by GBMiro — APPROVED** [review]

File: Penteract/Assets/Scripts/AbilityRefreshEffect.cpp (hunk 1 above)
Comment (devildrake) [reply Penteract/Assets/Scripts/AbilityRefreshEffect.cpp:66]: Yup

File: Penteract/Assets/Scripts/HUDController.cpp (hunk 2 above)
Comment (devildrake) [reply Penteract/Assets/Scripts/HUDController.cpp:394]: True

**Review by jcarlos0305 — APPROVED** [review]

File: Penteract/Assets/Scripts/HUDController.cpp [hunk 3]
```diff
@@ -25,6 +26,7 @@ EXPOSE_MEMBERS(HUDController) {
 		MEMBER(MemberType::GAME_OBJECT_UID, scoreTextUID),
 		MEMBER(MemberType::GAME_OBJECT_UID, lowHealthWarningEffectUID),
 		MEMBER(MemberType::FLOAT, timeToFadeDurableHealthFeedbackInternal)
+		//MEMBER(MemberType::GAME_OBJECT_UID,)
```
Comment (magalenyo) [inline Penteract/Assets/Scripts/HUDController.cpp:29]: Remove comment

File: Penteract/Assets/Scripts/HUDController.cpp [hunk 4]
```diff
@@ -290,15 +307,16 @@ void HUDController::UpdateComponents() {
 
 	if (!fangSkillsMainCanvas || !onimaruSkillsMainCanvas || !fangSkillsSecondCanvas || !onimaruSkillsSecondCanvas) return;
 	if (!fangHealthMainCanvas || !onimaruHealthMainCanvas || !fangHealthSecondCanvas || !onimaruHealthSecondCanvas) return;
+	//if (!fangSkillCooldownEffectCanvas || !onimaruSkillCoolDownEffectCanvas)return;
```
Comment (magalenyo) [inline Penteract/Assets/Scripts/HUDController.cpp:310]: remove comment

File: Penteract/Assets/Scripts/HUDController.cpp [hunk 5]
```diff
@@ -315,23 +333,27 @@ void HUDController::UpdateCommonSkill() {
 	}
 }
 
-void HUDController::UpdateFangCooldowns(GameObject* fangSkillCanvas) {
-	std::vector<GameObject*> children = fangSkillCanvas->GetChildren();
+
+void HUDController::UpdateFangCooldowns(GameObject* fangSkillCanvas, bool isMain) {
+	std::vector<GameObject*> skills = fangSkillCanvas->GetChildren();
 	int skill = FANG_SKILL_1;
```
Comment (magalenyo) [inline Penteract/Assets/Scripts/HUDController.cpp:339]: !!!!!! Skill value never changes so it is always going to be calling AbilityCooldownEffectCheck(1..)

File: Penteract/Assets/Scripts/HUDController.h [hunk 6]
```diff
@@ -117,6 +122,7 @@ class HUDController : public Script {
 
 	float remainingTimesFang[MAX_HEALTH] = { 0,0,0,0,0,0,0 };
 	float remainingTimesOni[MAX_HEALTH] = { 0,0,0,0,0,0,0 };
+	bool abilityCoolDownsRetreived[Cooldowns::TOTAL] = { true,true,true,true,true,true,true };
```
Comment (magalenyo) [inline Penteract/Assets/Scripts/HUDController.h:125]: I couldn't comment on the declaration of the Enum Cooldowns, but if it isn't used somewhere else, have it as private enum

File: Penteract/Assets/Scripts/HUDController.cpp [hunk 7]
```diff
@@ -315,23 +333,27 @@ void HUDController::UpdateCommonSkill() {
...
-		std::vector<GameObject*> skills = (*it)->GetChildren();
-		for (std::vector<GameObject*>::iterator itSkills = skills.begin(); itSkills != skills.end(); ++itSkills) {
-			ComponentImage* image = (*itSkills)->GetComponent<ComponentImage>();
+	for (std::vector<GameObject*>::iterator it = skills.begin(); it != skills.end(); ++it) {
+
+		if ((*it)->GetChildren().size() > 0) {
+			ComponentImage* image = (*it)->GetChildren()[1]->GetComponent<ComponentImage>();
 			if (image) {
 				if (image->IsFill()) {
 					switch (skill) {
```
Comment (magalenyo) [inline Penteract/Assets/Scripts/HUDController.cpp:346]: There's no need for the switch. You can simply go: 

AbilityCoolDownEffectCheck(skill, fangSkillCanvas);
SetFillValue(cooldowns[skill]);

File: Penteract/Assets/Scripts/HUDController.cpp [hunk 8]
```diff
@@ -340,11 +362,88 @@ void HUDController::UpdateFangCooldowns(GameObject* fangSkillCanvas) {
...
+
+				PlayCoolDownEffect(ef, cooldown);
+			}
+		}
+	}
+}
+
+
+
+void HUDController::UpdateOnimaruCooldowns(GameObject* onimaruSkillCanvas, bool isMain) {
```
Comment (magalenyo) [inline Penteract/Assets/Scripts/HUDController.cpp:446]: isn't this the same function as Fang? Can't it be merged so you don't have two separate functions that appear to do the same?

File: Penteract/Assets/Scripts/HUDController.cpp [hunk 9]
```diff
@@ -340,11 +362,88 @@ void HUDController::UpdateFangCooldowns(GameObject* fangSkillCanvas) {
...
+}
+
+
+void HUDController::AbilityCoolDownEffectCheck(Cooldowns cooldown, GameObject* canvas) {
+	if (!abilityCoolDownsRetreived[cooldown]) {
+		if (canvas) {
+			if (cooldowns[cooldown] == 1.0f) {
+
+				AbilityRefreshEffect* ef = nullptr;
+				switch (cooldown) {
```
Comment (magalenyo) [inline Penteract/Assets/Scripts/HUDController.cpp:386]: You don't need this switch case. Every case does the same except for the index of the GetChildren. Might be better to have the size as index and then do a generic if to check if it has a children. Same for Onimaru.

**Review by magalenyo — APPROVED** [review]
None of my comments are necessary and the work is fine, but I think there could be an improvement in quality:
- There are many new scenes, and I don't know if they are necessary . I'd suggest removing the unnecessary ones, and if needed, merge them into a single scene to test.
- There's plenty of repeated code which could be reduced into a new function, or in a couple of generic statements.

I haven't tested it but I'm sure this works fine. Again, **PR approved**, you don't need to follow the comments! :)

File: Penteract/Assets/Scripts/HUDController.cpp (hunk 3 above)
Comment (devildrake) [reply Penteract/Assets/Scripts/HUDController.cpp:29]: Yup

File: Penteract/Assets/Scripts/HUDController.cpp (hunk 4 above)
Comment (devildrake) [reply Penteract/Assets/Scripts/HUDController.cpp:310]: Yupp

File: Penteract/Assets/Scripts/HUDController.cpp (hunk 5 above)
Comment (devildrake) [reply Penteract/Assets/Scripts/HUDController.cpp:339]: It does change in line 365!

File: Penteract/Assets/Scripts/HUDController.cpp (hunk 9 above)
Comment (devildrake) [reply Penteract/Assets/Scripts/HUDController.cpp:386]: More generic version implemented.

File: Penteract/Assets/Scripts/HUDController.cpp (hunk 8 above)
Comment (devildrake) [reply Penteract/Assets/Scripts/HUDController.cpp:446]: We may need to have a meet about this one.

File: Penteract/Assets/Scripts/HUDController.cpp (hunk 7 above)
Comment (devildrake) [reply Penteract/Assets/Scripts/HUDController.cpp:346]: Yup

File: Penteract/Assets/Scripts/HUDController.h (hunk 6 above)
Comment (devildrake) [reply Penteract/Assets/Scripts/HUDController.h:125]: Changed to enum class on Jaime's orders, forward declaration of enum appears troublesome

Comment (magalenyo) [reply Penteract/Assets/Scripts/HUDController.h]: What I meant is to have the declaration of the enum inside the private scope, not outside the class.

**Pull request merged**